Decouple peer-processing-logic from block-connection-logic + BIP61 removal.

furszy opened the pull request:
Continuation of #2463. Another error prone topic.

Essentially focused in the following points:
* Remove duplicate and incorrect peer misbehaving:
-1) The peer misbehaving score set for an invalid arriving block is being performed inside `PeerLogicValidation::BlockChecked`.
-2) The `CValidationState` param in `ProcessNewBlock` does not return the error nor the invalidity reason if the block is marked invalid during the activate best chain process (block connection), an empty `CValidationState` is set. The `BlockChecked` signal is used instead to notify the block invalidity reason.

* Get rid off `CValidationState` param in `ProcessNewBlock` and use only `BlockChecked` (own version of ae223576077448bd4ec250c73f5d8fe5e9a9ac7d)
* Removal of BIP61 (#15437 and #18609).
-- BIP61 could be disabled first by default before its removal like upstream did (or moved to its own standalone PR), but.. i don't think that worth to continue maintaining it, most likely no one is using it and if someone is actually using it, isn't good to have any piece of software depending on it. --
* Move block spam filter check to `BlockChecked`. Following the same peer-processing-logic / block-connection-logic division. Improves the spam protection as now blocks rejected during the connection process are going to pass through the spam filter check as well (before, only blocks rejected during `AcceptBlock` were passing through it).
Plus, this improvement let me do some further cleanup and remove the extra `fAccepted` flag from `ProcessNewBlock`.
* Create `GetDataMsg` enum and replace `ppszTypeName` array for direct `GetCommand` call.
The array was duplicated with the `NetMsgType` constants

furszy commented:
Rebased on master, lint-python warnings fixed. Ready to go.

random-zebra commented on src/util/blockstatecatcher.h:
```diff
@@ -0,0 +1,35 @@
+    bool stateErrorFound() { return found && state.IsError(); }
+
+protected:
+    virtual void BlockChecked(const CBlock& block, const CValidationState& stateIn) {
+        if (block.GetHash() != hash) return; found = true; state = stateIn;
+    };
```
please keep this on 3 or 4 lines (as it was in `submitblock_StateCatcher`), otherwise GCC complains about "misleading indentation".
```
util/blockstatecatcher.h:31:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if (block.GetHash() != hash) return; found = true; state = stateIn;
         ^~
util/blockstatecatcher.h:31:46: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
         if (block.GetHash() != hash) return; found = true; state = stateIn;
                                              ^~~~~
```

random-zebra reviewed (commented):
Really nice PR. Just few nits

furszy commented on doc/release-notes.md:
```diff
@@ -493,6 +493,36 @@ OpenSSL is no longer used by PIVX Core
+
+* Testing the validity of a transaction can be achieved by specific RPCs:
+  - `sendrawtransaction`
+  - `testmempoolaccept`
```
you know, was absolutely sure that i already back ported it but seems that not.

furszy commented on src/net_processing.cpp:
```diff
@@ -1899,51 +1882,29 @@ bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vR
-                found = true;
-                break;
-            }
+    // Tier two msg type search
```
yes ☕

furszy replied:
Added it inside f16f4224515ad0468c54795ccf055e8743c4858d with further improvements.

furszy commented on src/test/util/blocksutil.cpp:
```diff
@@ -4,39 +4,20 @@
     RegisterValidationInterface(&stateChecker);
-    ProcessNewBlock(state, pblock, nullptr);
+    ProcessNewBlock(pblock, nullptr);
     UnregisterValidationInterface(&stateChecker);
```
yeah, this was the first implementation that made. Updating..

furszy commented:
Updated per zebra's feedback, squashing all of the updates in their respective commits, and rebased it on master so it has the correct bench raw files generation.

To minimize the review mental burden --> `git range-diff master e1c01c4 f16f4224`

random-zebra commented on src/protocol.cpp:
```diff
@@ -226,3 +226,27 @@ const std::vector<std::string>& getAllNetMessageTypes()
     return allNetMessageTypesVec;
 }
+
+const std::vector<std::string>& getTierTwoNetMessageTypes()
+{
+    // Note: keep this updated with allNetMessageTypes
+    const static std::vector<std::string> allTierTwoMessageTypesVec {
+            NetMsgType::SPORK,
+            NetMsgType::MNWINNER,
+            "mnodescanerr",
+            NetMsgType::BUDGETVOTE,
+            NetMsgType::BUDGETPROPOSAL,
+            NetMsgType::FINALBUDGET,
+            NetMsgType::FINALBUDGETVOTE,
+            "mnq",
+            NetMsgType::MNBROADCAST,
+            NetMsgType::MNPING,
+            "dstx",  // deprecated
+            NetMsgType::GETMNWINNERS,
+            NetMsgType::GETMNLIST,
+            NetMsgType::BUDGETVOTESYNC,
+            NetMsgType::GETSPORKS,
+            NetMsgType::SYNCSTATUSCOUNT
+    };
+    return allTierTwoMessageTypesVec;
+}
```
I'm not a fan of this kind of duplication, where two static constant lists depend on each other, and, when we modify one we have to remember to update the other one too.
It's easy to miss the comments and introduce bugs this way (it happened to me in the past).

I propose this alternative:
```patch
Index: src/protocol.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/protocol.cpp b/src/protocol.cpp
--- a/src/protocol.cpp	(revision f16f4224515ad0468c54795ccf055e8743c4858d)
+++ b/src/protocol.cpp	(date 1626951151800)
@@ -80,7 +80,7 @@
     "filtered block", // Should never occur
     "ix",   // deprecated
     "txlvote", // deprecated
-    NetMsgType::SPORK,
+    NetMsgType::SPORK,           // --- tiertwoNetMessageTypes start here ---
     NetMsgType::MNWINNER,
     "mnodescanerr",
     NetMsgType::BUDGETVOTE,
@@ -98,6 +98,7 @@
     NetMsgType::SYNCSTATUSCOUNT
 };
 const static std::vector<std::string> allNetMessageTypesVec(allNetMessageTypes, allNetMessageTypes + ARRAYLEN(allNetMessageTypes));
+const static std::vector<std::string> tiertwoNetMessageTypesVec(std::find(allNetMessageTypesVec.begin(), allNetMessageTypesVec.end(), NetMsgType::SPORK), allNetMessageTypesVec.end());
 
 CMessageHeader::CMessageHeader(const MessageStartChars& pchMessageStartIn)
 {
@@ -229,24 +230,5 @@
 
 const std::vector<std::string>& getTierTwoNetMessageTypes()
 {
-    // Note: keep this updated with allNetMessageTypes
-    const static std::vector<std::string> allTierTwoMessageTypesVec {
-            NetMsgType::SPORK,
-            NetMsgType::MNWINNER,
-            "mnodescanerr",
-            NetMsgType::BUDGETVOTE,
-            NetMsgType::BUDGETPROPOSAL,
-            NetMsgType::FINALBUDGET,
-            NetMsgType::FINALBUDGETVOTE,
-            "mnq",
-            NetMsgType::MNBROADCAST,
-            NetMsgType::MNPING,
-            "dstx",  // deprecated
-            NetMsgType::GETMNWINNERS,
-            NetMsgType::GETMNLIST,
-            NetMsgType::BUDGETVOTESYNC,
-            NetMsgType::GETSPORKS,
-            NetMsgType::SYNCSTATUSCOUNT
-    };
-    return allTierTwoMessageTypesVec;
-}
\ No newline at end of file
+    return tiertwoNetMessageTypesVec;
+}
```

furszy replied:
sounds good, it will still need a good comment to modify the vector in case of adding non tier two related messages at the end of the vector.

furszy commented:
updated, added the vector change.

random-zebra reviewed (approved):
ACK 8e9f2bcce6c3adb542a36a55d4734aa8c1329f7e
